File: node-backend/src/helpers/module_A/pageMetrics/wordCountExtractor.ts
import type { CheerioAPI } from 'cheerio';
import crypto from 'crypto';

/**
 * Word Count Data for Page Metrics
 */
export interface PageWordCountData {
    totalWordCount: number; // Total word count (all text, excluding script/style/noscript)
    visibleWordCount: number; // Visible word count (main content only)
    uniqueWordCount: number;
    textToHtmlRatio: number;
    sentenceCount: number;
    paragraphCount: number;
    averageSentenceLength: number;
    averageParagraphLength: number;
    keywordDensity: number | null; // Keyword density percentage (0-100) or null if no keyword provided
    thinContent: boolean; // Whether page has thin content
    thinContentReason: string | null; // Reason: 'Low word count' or 'Low uniqueness'
    duplicateContent: boolean; // Whether page content is duplicate (requires session context)
    duplicateWithUrls: string[]; // URLs with duplicate content (requires session context)
    sectionWordCountMapping: Record<string, number>; // Section heading -> word count
    sectionWordCountBreakdown: Record<string, number>; // Section heading -> percentage
    headingWordCountMapping: Record<string, number>; // Heading text -> word count under heading
}

/**
 * Extract visible text from DOM, ignoring hidden elements
 */
function extractVisibleText($: CheerioAPI): string {
    const $clone = $.load($.html());
    
    // Remove script, style, noscript tags
    $clone('script').remove();
    $clone('style').remove();
    $clone('noscript').remove();
    
    // Remove elements with display:none, visibility:hidden, or aria-hidden="true"
    $clone('*').each((_i, el) => {
        const $el = $clone(el);
        const style = $el.attr('style') || '';
        const ariaHidden = $el.attr('aria-hidden');
        
        // Check for display:none or visibility:hidden in inline styles
        if (style.includes('display:none') || style.includes('display: none') || 
            style.includes('visibility:hidden') || style.includes('visibility: hidden') ||
            ariaHidden === 'true') {
            $el.remove();
        }
    });
    
    // Optionally remove footer, nav, sidebar (common boilerplate)
    $clone('footer, nav, aside').remove();
    
    // Get visible text from body
    return $clone('body').text() || '';
}

/**
 * Normalize text for word counting (lowercase, remove punctuation)
 */
function normalizeText(text: string): string {
    return text
        .toLowerCase()
        .replace(/[^\w\s]/g, ' ') // Replace punctuation with spaces
        .replace(/\s+/g, ' ') // Normalize whitespace
        .trim();
}

/**
 * Tokenize text into words
 */
function tokenizeWords(text: string): string[] {
    return text
        .split(/\s+/)
        .filter(word => word.trim().length > 0);
}

/**
 * Extract total word count from HTML page (all text, excluding script/style/noscript)
 * Algorithm:
 * 1. Remove script, style, noscript tags
 * 2. Extract all remaining text
 * 3. Tokenize and count
 */
function extractTotalWordCount($: CheerioAPI): number {
    const $clone = $.load($.html());
    
    // Remove script, style, and noscript tags
    $clone('script').remove();
    $clone('style').remove();
    $clone('noscript').remove();
    
    // Get all text content
    const textContent = $clone('body').text() || '';
    const words = tokenizeWords(textContent);
    return words.length;
}

/**
 * Extract visible word count from HTML page
 * Algorithm:
 * 1. Parse DOM
 * 2. Ignore: display:none, visibility:hidden, aria-hidden="true", Footer, nav, sidebar
 * 3. Extract text from visible nodes
 * 4. Tokenize and count
 */
function extractVisibleWordCount($: CheerioAPI): number {
    const visibleText = extractVisibleText($);
    const words = tokenizeWords(visibleText);
    return words.length;
}

/**
 * Extract unique word count
 * Algorithm:
 * 1. Take visible words
 * 2. Normalize: lowercase, remove punctuation
 * 3. Convert to a set
 * 4. Count unique items
 */
function extractUniqueWordCount($: CheerioAPI): number {
    const visibleText = extractVisibleText($);
    const normalizedText = normalizeText(visibleText);
    const words = tokenizeWords(normalizedText);
    const uniqueWords = new Set(words);
    return uniqueWords.size;
}

/**
 * Calculate text-to-HTML ratio
 * Algorithm:
 * 1. Measure total HTML size (bytes)
 * 2. Measure visible text size (bytes)
 * 3. Calculate ratio
 */
function calculateTextToHtmlRatio($: CheerioAPI): number {
    const html = $.html();
    const htmlSizeBytes = Buffer.byteLength(html, 'utf8');
    
    const visibleText = extractVisibleText($);
    const textSizeBytes = Buffer.byteLength(visibleText, 'utf8');
    
    if (htmlSizeBytes === 0) return 0;
    
    const ratio = (textSizeBytes / htmlSizeBytes) * 100;
    return Math.round(ratio * 100) / 100; // Round to 2 decimal places
}

/**
 * Extract sentence count
 * Algorithm:
 * 1. Take visible text
 * 2. Split on sentence delimiters: ., !, ?
 * 3. Filter empty results
 * 4. Count
 */
function extractSentenceCount($: CheerioAPI): number {
    const visibleText = extractVisibleText($);
    
    if (!visibleText.trim()) return 0;
    
    // Split on sentence delimiters
    const sentences = visibleText
        .split(/[.!?]+/)
        .map(s => s.trim())
        .filter(s => s.length > 0);
    
    // If no sentences found but we have text, treat entire text as one sentence
    if (sentences.length === 0 && visibleText.trim().length > 0) {
        const words = tokenizeWords(visibleText);
        return words.length > 0 ? 1 : 0;
    }
    
    return sentences.length;
}

/**
 * Extract paragraph count
 * Algorithm:
 * 1. Count <p> tags with text
 * 2. Optionally include <div> blocks with large text
 * 3. Ignore empty paragraphs
 */
function extractParagraphCount($: CheerioAPI): number {
    const $clone = $.load($.html());
    
    // Remove hidden elements first
    $clone('script, style, noscript').remove();
    $clone('*').each((_i, el) => {
        const $el = $clone(el);
        const style = $el.attr('style') || '';
        const ariaHidden = $el.attr('aria-hidden');
        
        if (style.includes('display:none') || style.includes('display: none') || 
            style.includes('visibility:hidden') || style.includes('visibility: hidden') ||
            ariaHidden === 'true') {
            $el.remove();
        }
    });
    
    // Count <p> tags with non-empty text
    let paragraphCount = 0;
    $clone('p').each((_i, el) => {
        const text = $clone(el).text().trim();
        if (text.length > 0) {
            paragraphCount++;
        }
    });
    
    // If no <p> tags, count <div> blocks with substantial text (optional)
    if (paragraphCount === 0) {
        $clone('div').each((_i, el) => {
            const text = $clone(el).text().trim();
            const wordCount = tokenizeWords(text).length;
            // Consider divs with at least 20 words as paragraphs
            if (wordCount >= 20) {
                paragraphCount++;
            }
        });
    }
    
    // At least 1 if we have any visible text
    if (paragraphCount === 0) {
        const visibleText = extractVisibleText($);
        if (visibleText.trim().length > 0) {
            paragraphCount = 1;
        }
    }
    
    return paragraphCount;
}

/**
 * Calculate average sentence length
 * Algorithm: visible_word_count / sentence_count
 */
function calculateAverageSentenceLength(visibleWordCount: number, sentenceCount: number): number {
    if (sentenceCount === 0) return 0;
    return Math.round((visibleWordCount / sentenceCount) * 100) / 100; // Round to 2 decimal places
}

/**
 * Calculate average paragraph length
 * Algorithm: visible_word_count / paragraph_count
 */
function calculateAverageParagraphLength(visibleWordCount: number, paragraphCount: number): number {
    if (paragraphCount === 0) return 0;
    return Math.round((visibleWordCount / paragraphCount) * 100) / 100; // Round to 2 decimal places
}

/**
 * Calculate keyword density with proper algorithm
 * Algorithm:
 * 1. Extract visible text from page
 * 2. Normalize text: lowercase, remove punctuation, normalize whitespace
 * 3. Tokenize into words array
 * 4. Normalize keyword: lowercase, remove punctuation
 * 5. For single-word keywords: count exact word matches
 * 6. For multi-word keywords: count phrase occurrences (consecutive words)
 * 7. Calculate density: (keyword_occurrences / total_visible_words) * 100
 * 
 * @param $ - Cheerio instance with loaded HTML
 * @param targetKeyword - Target keyword or phrase to calculate density for
 * @returns Keyword density percentage (0-100) or null if invalid
 */
function calculateKeywordDensity($: CheerioAPI, targetKeyword?: string): number | null {
    // Validate input
    if (!targetKeyword || typeof targetKeyword !== 'string') {
        return null;
    }
    
    const trimmedKeyword = targetKeyword.trim();
    if (trimmedKeyword.length === 0) {
        return null;
    }
    
    // Extract visible text
    const visibleText = extractVisibleText($);
    if (!visibleText || visibleText.trim().length === 0) {
        return null;
    }
    
    // Normalize text: lowercase, remove punctuation, normalize whitespace
    const normalizedText = normalizeText(visibleText);
    if (normalizedText.length === 0) {
        return null;
    }
    
    // Tokenize into words array
    const words = tokenizeWords(normalizedText);
    const totalWords = words.length;
    
    if (totalWords === 0) {
        return null;
    }
    
    // Normalize keyword: lowercase, remove punctuation, normalize whitespace
    const normalizedKeyword = normalizeText(trimmedKeyword);
    if (normalizedKeyword.length === 0) {
        return null;
    }
    
    // Tokenize keyword into words
    const keywordWords = tokenizeWords(normalizedKeyword);
    if (keywordWords.length === 0) {
        return null;
    }
    
    let occurrences = 0;
    
    if (keywordWords.length === 1) {
        // Single-word keyword: count exact word matches
        const keyword = keywordWords[0];
        occurrences = words.filter(word => word === keyword).length;
    } else {
        // Multi-word keyword phrase: count consecutive phrase occurrences using sliding window
        // This ensures we match the exact phrase sequence, not just individual words
        const keywordLength = keywordWords.length;
        
        // Slide a window through the words array to find matching phrases
        for (let i = 0; i <= words.length - keywordLength; i++) {
            // Check if the next N words match the keyword phrase
            let matches = true;
            for (let j = 0; j < keywordLength; j++) {
                if (words[i + j] !== keywordWords[j]) {
                    matches = false;
                    break;
                }
            }
            if (matches) {
                occurrences++;
                // Skip ahead to avoid overlapping matches (optional - remove if you want overlapping)
                // i += keywordLength - 1;
            }
        }
    }
    
    if (occurrences === 0) {
        return 0; // Return 0 instead of null if keyword not found
    }
    
    // Calculate density: (occurrences / total_words) * 100
    const density = (occurrences / totalWords) * 100;
    
    // Round to 2 decimal places
    return Math.round(density * 100) / 100;
}

/**
 * Detect thin content based on word count and uniqueness
 * Algorithm:
 * 1. Check if visible_word_count < 250
 * 2. Check if unique_word_count / visible_word_count < 0.3
 * 3. Return thin_content flag and reason
 */
function detectThinContent(visibleWordCount: number, uniqueWordCount: number): { thinContent: boolean; reason: string | null } {
    if (visibleWordCount < 250) {
        return { thinContent: true, reason: 'Low word count' };
    }
    
    if (visibleWordCount > 0) {
        const uniquenessRatio = uniqueWordCount / visibleWordCount;
        if (uniquenessRatio < 0.3) {
            return { thinContent: true, reason: 'Low uniqueness' };
        }
    }
    
    return { thinContent: false, reason: null };
}

/**
 * Normalize content for duplicate detection
 * Algorithm:
 * 1. Lowercase
 * 2. Remove boilerplate (header, footer, nav)
 * 3. Normalize whitespace
 */
function normalizeContentForDuplicate(visibleText: string): string {
    return visibleText
        .toLowerCase()
        .replace(/\s+/g, ' ')
        .trim();
}

/**
 * Generate content hash for duplicate detection
 * Uses MD5 hash of normalized content
 */
function generateContentHashForDuplicate(visibleText: string): string {
    const normalized = normalizeContentForDuplicate(visibleText);
    return crypto.createHash('md5').update(normalized).digest('hex');
}

/**
 * Extract section word count mapping
 * Algorithm:
 * 1. Identify sections (each section starts at a heading <h1>-<h6>)
 * 2. Collect text from heading until next heading of same or higher level
 * 3. Count words per section
 */
function extractSectionWordCountMapping($: CheerioAPI): Record<string, number> {
    const $clone = $.load($.html());
    
    // Remove hidden elements
    $clone('script, style, noscript').remove();
    $clone('*').each((_i, el) => {
        const $el = $clone(el);
        const style = $el.attr('style') || '';
        const ariaHidden = $el.attr('aria-hidden');
        if (style.includes('display:none') || style.includes('display: none') || 
            style.includes('visibility:hidden') || style.includes('visibility: hidden') ||
            ariaHidden === 'true') {
            $el.remove();
        }
    });
    
    // Remove footer, nav, aside
    $clone('footer, nav, aside').remove();
    
    const sectionMapping: Record<string, number> = {};
    const headings = $clone('h1, h2, h3, h4, h5, h6').toArray();
    
    if (headings.length === 0) {
        // No headings, treat entire body as one section
        const bodyText = $clone('body').text() || '';
        const wordCount = tokenizeWords(bodyText).length;
        if (wordCount > 0) {
            sectionMapping['Main Content'] = wordCount;
        }
        return sectionMapping;
    }
    
    // Process each heading
    for (let i = 0; i < headings.length; i++) {
        const heading = headings[i];
        const $heading = $clone(heading);
        const headingLevel = parseInt($heading.prop('tagName')?.replace('H', '') || '0');
        const headingText = $heading.text().trim();
        
        if (!headingText) continue;
        
        // Collect text from this heading until next heading of same or higher level
        let sectionText = headingText + ' ';
        let current = $heading.next();
        
        while (current.length > 0) {
            const tagName = current.prop('tagName')?.toUpperCase() || '';
            
            // Check if we hit another heading
            if (tagName.match(/^H[1-6]$/)) {
                const nextHeadingLevel = parseInt(tagName.replace('H', ''));
                // Stop if we hit a heading of same or higher level
                if (nextHeadingLevel <= headingLevel) {
                    break;
                }
            }
            
            // Add text content
            const text = current.text().trim();
            if (text) {
                sectionText += text + ' ';
            }
            
            current = current.next();
        }
        
        // Count words in this section
        const wordCount = tokenizeWords(sectionText).length;
        if (wordCount > 0) {
            sectionMapping[headingText] = wordCount;
        }
    }
    
    return sectionMapping;
}

/**
 * Calculate section word count breakdown (percentage distribution)
 * Algorithm:
 * 1. Get section word count mapping
 * 2. Calculate total words across all sections
 * 3. Calculate percentage for each section
 */
function calculateSectionWordCountBreakdown(
    sectionMapping: Record<string, number>,
    totalVisibleWords: number
): Record<string, number> {
    if (totalVisibleWords === 0) {
        return {};
    }
    
    const breakdown: Record<string, number> = {};
    
    for (const [section, wordCount] of Object.entries(sectionMapping)) {
        const percentage = (wordCount / totalVisibleWords) * 100;
        breakdown[section] = Math.round(percentage * 100) / 100; // Round to 2 decimal places
    }
    
    return breakdown;
}

/**
 * Extract heading word count mapping
 * Algorithm:
 * 1. Extract all headings (H1-H6)
 * 2. For each heading, collect content until next relevant heading
 * 3. Count words under each heading
 */
function extractHeadingWordCountMapping($: CheerioAPI): Record<string, number> {
    const $clone = $.load($.html());
    
    // Remove hidden elements
    $clone('script, style, noscript').remove();
    $clone('*').each((_i, el) => {
        const $el = $clone(el);
        const style = $el.attr('style') || '';
        const ariaHidden = $el.attr('aria-hidden');
        if (style.includes('display:none') || style.includes('display: none') || 
            style.includes('visibility:hidden') || style.includes('visibility: hidden') ||
            ariaHidden === 'true') {
            $el.remove();
        }
    });
    
    // Remove footer, nav, aside
    $clone('footer, nav, aside').remove();
    
    const headingMapping: Record<string, number> = {};
    const headings = $clone('h1, h2, h3, h4, h5, h6').toArray();
    
    for (let i = 0; i < headings.length; i++) {
        const heading = headings[i];
        const $heading = $clone(heading);
        const headingLevel = parseInt($heading.prop('tagName')?.replace('H', '') || '0');
        const headingText = $heading.text().trim();
        
        if (!headingText) continue;
        
        // Create heading key with level prefix
        const headingKey = `H${headingLevel}: ${headingText}`;
        
        // Collect text from this heading until next heading of same or higher level
        let sectionText = '';
        let current = $heading.next();
        
        while (current.length > 0) {
            const tagName = current.prop('tagName')?.toUpperCase() || '';
            
            // Check if we hit another heading
            if (tagName.match(/^H[1-6]$/)) {
                const nextHeadingLevel = parseInt(tagName.replace('H', ''));
                // Stop if we hit a heading of same or higher level
                if (nextHeadingLevel <= headingLevel) {
                    break;
                }
            }
            
            // Add text content
            const text = current.text().trim();
            if (text) {
                sectionText += text + ' ';
            }
            
            current = current.next();
        }
        
        // Count words under this heading
        const wordCount = tokenizeWords(sectionText).length;
        headingMapping[headingKey] = wordCount;
    }
    
    return headingMapping;
}

/**
 * Extract comprehensive word count analysis from HTML page
 * Implements all wordcount analysis algorithms as specified
 * 
 * @param $ - Cheerio instance with loaded HTML
 * @param targetKeyword - Optional target keyword for density calculation
 * @returns PageWordCountData with all metrics
 */
export function extractWordCount($: CheerioAPI, targetKeyword?: string): PageWordCountData {
    // Extract total word count (all text, excluding script/style/noscript)
    const totalWordCount = extractTotalWordCount($);
    
    // Extract visible word count
    const visibleWordCount = extractVisibleWordCount($);
    
    // Extract unique word count
    const uniqueWordCount = extractUniqueWordCount($);
    
    // Calculate text-to-HTML ratio
    const textToHtmlRatio = calculateTextToHtmlRatio($);
    
    // Extract sentence count
    const sentenceCount = extractSentenceCount($);
    
    // Extract paragraph count
    const paragraphCount = extractParagraphCount($);
    
    // Calculate average sentence length
    const averageSentenceLength = calculateAverageSentenceLength(visibleWordCount, sentenceCount);
    
    // Calculate average paragraph length
    const averageParagraphLength = calculateAverageParagraphLength(visibleWordCount, paragraphCount);
    
    // Calculate keyword density (if target keyword provided)
    const keywordDensity = calculateKeywordDensity($, targetKeyword);
    
    // Detect thin content
    const { thinContent, reason: thinContentReason } = detectThinContent(visibleWordCount, uniqueWordCount);
    
    // Extract section word count mapping
    const sectionWordCountMapping = extractSectionWordCountMapping($);
    
    // Calculate section word count breakdown
    const sectionWordCountBreakdown = calculateSectionWordCountBreakdown(sectionWordCountMapping, visibleWordCount);
    
    // Extract heading word count mapping
    const headingWordCountMapping = extractHeadingWordCountMapping($);
    
    // Duplicate content detection requires session context (will be done separately)
    // For now, set defaults
    const duplicateContent = false;
    const duplicateWithUrls: string[] = [];
    
    // Ensure totalWordCount is at least equal to visibleWordCount (sanity check)
    const finalTotalWordCount = totalWordCount >= visibleWordCount ? totalWordCount : visibleWordCount;
    
    return {
        totalWordCount: finalTotalWordCount,
        visibleWordCount,
        uniqueWordCount,
        textToHtmlRatio,
        sentenceCount,
        paragraphCount,
        averageSentenceLength,
        averageParagraphLength,
        keywordDensity,
        thinContent,
        thinContentReason,
        duplicateContent,
        duplicateWithUrls,
        sectionWordCountMapping,
        sectionWordCountBreakdown,
        headingWordCountMapping
    };
}

/**
 * Generate content hash for duplicate detection
 * This can be used to compare with other pages in the session
 */
export function generateContentHash(visibleText: string): string {
    return generateContentHashForDuplicate(visibleText);
}

File: npy-backend/modules/module_A/Wordcount_analysis/wordcount_extractor.py
import re
import hashlib
from typing import Dict, List, Any, Optional, Set
from bs4 import BeautifulSoup, Comment
from urllib.parse import urlparse

def extract_visible_text(soup: BeautifulSoup) -> str:
    """
    Extract visible text from BeautifulSoup object, ignoring hidden elements, scripts, styles.
    """
    # Create a copy to avoid mutating original
    temp_soup = BeautifulSoup(str(soup), 'html.parser')
    
    # Remove script, style, noscript, etc.
    for element in temp_soup(["script", "style", "noscript", "header", "footer", "nav", "aside"]):
        element.decompose()
        
    # Remove elements with hidden styles in style attribute
    for element in temp_soup.find_all(style=True):
        style = element.get('style', '').lower()
        if 'display:none' in style or 'display: none' in style or \
           'visibility:hidden' in style or 'visibility: hidden' in style:
            element.decompose()
            
    # Remove elements with aria-hidden="true"
    for element in temp_soup.find_all(attrs={"aria-hidden": "true"}):
        element.decompose()

    # Get text
    text = temp_soup.get_text(separator=' ')
    
    # Clean whitespace
    lines = (line.strip() for line in text.splitlines())
    chunks = (phrase.strip() for line in lines for phrase in line.split("  "))
    text = ' '.join(chunk for chunk in chunks if chunk)
    
    return text

def normalize_text(text: str) -> str:
    """
    Normalize text for word counting (lowercase, remove punctuation).
    """
    # Replace non-alphanumeric with spaces
    text = re.sub(r'[^\w\s]', ' ', text.lower())
    # Normalize whitespace
    text = re.sub(r'\s+', ' ', text).strip()
    return text

def tokenize_words(text: str) -> List[str]:
    """
    Tokenize text into words.
    """
    return [w for w in text.split() if w.strip()]

def get_sentence_count(text: str) -> int:
    """
    Extract sentence count using delimiters.
    """
    if not text.strip():
        return 0
    sentences = re.split(r'[.!?]+', text)
    sentences = [s.strip() for s in sentences if s.strip()]
    if not sentences and text.strip():
        return 1
    return len(sentences)

def get_paragraph_count(soup: BeautifulSoup) -> int:
    """
    Count paragraphs with substantial text.
    """
    p_tags = soup.find_all('p')
    count = 0
    for p in p_tags:
        if p.get_text().strip():
            count += 1
            
    if count == 0:
        # Fallback to divs with substantial text
        div_tags = soup.find_all('div')
        for div in div_tags:
            text = div.get_text().strip()
            if len(tokenize_words(text)) >= 20:
                count += 1
                
    if count == 0:
        visible_text = extract_visible_text(soup)
        if visible_text.strip():
            count = 1
            
    return count

def calculate_keyword_density(text: str, keyword: Optional[str]) -> Optional[float]:
    """
    Calculate keyword density (0-100).
    """
    if not keyword or not keyword.strip():
        return None
    
    norm_text = normalize_text(text)
    words = tokenize_words(norm_text)
    if not words:
        return 0.0
    
    norm_keyword = normalize_text(keyword)
    keyword_words = tokenize_words(norm_keyword)
    if not keyword_words:
        return 0.0
    
    count = 0
    if len(keyword_words) == 1:
        target = keyword_words[0]
        count = words.count(target)
    else:
        # Multi-word sliding window
        target_len = len(keyword_words)
        for i in range(len(words) - target_len + 1):
            if words[i:i+target_len] == keyword_words:
                count += 1
                
    density = (count / len(words)) * 100
    return round(density, 2)

def detect_thin_content(visible_word_count: int, unique_word_count: int) -> Dict[str, Any]:
    """
    Detect thin content based on word count and uniqueness.
    """
    if visible_word_count < 250:
        return {"thinContent": True, "thinContentReason": "Low word count"}
    
    if visible_word_count > 0:
        uniqueness_ratio = unique_word_count / visible_word_count
        if uniqueness_ratio < 0.3:
            return {"thinContent": True, "thinContentReason": "Low uniqueness"}
            
    return {"thinContent": False, "thinContentReason": None}

def extract_section_mapping(soup: BeautifulSoup) -> Dict[str, int]:
    """
    Map words to specific sections identified by headers.
    """
    # Simple implementation: find headers and collect text until next header
    mapping = {}
    headers = soup.find_all(['h1', 'h2', 'h3', 'h4', 'h5', 'h6'])
    
    if not headers:
        text = extract_visible_text(soup)
        wc = len(tokenize_words(text))
        if wc > 0:
            mapping["Main Content"] = wc
        return mapping

    # Process each header
    for i, h in enumerate(headers):
        h_text = h.get_text().strip()
        if not h_text:
            continue
        
        # Collect text from next siblings until next header
        section_text = [h_text]
        for sibling in h.next_siblings:
            if sibling.name in ['h1', 'h2', 'h3', 'h4', 'h5', 'h6']:
                break
            if hasattr(sibling, 'get_text'):
                section_text.append(sibling.get_text())
                
        wc = len(tokenize_words(' '.join(section_text)))
        mapping[h_text] = wc
        
    return mapping

def generate_content_hash(text: str) -> str:
    """
    Generates an MD5 hash of normalized content.
    """
    norm = re.sub(r'\s+', ' ', text.lower()).strip()
    return hashlib.md5(norm.encode('utf-8')).hexdigest()

def detect_intent(visible_text: str, url: str) -> str:
    """
    Simple intent detection: Informational (Blog) vs Transactional (Product).
    Ported from SergeService.ts detectIntent logic.
    """
    text_lower = visible_text.lower()
    url_lower = url.lower()
    
    blog_indicators = ['blog', 'guide', 'how to', 'tips', 'what is', 'learn', 'explained', 'definition']
    product_indicators = ['product', 'pricing', 'plans', 'buy', 'order', 'shop', 'deal', 'coupon', 'price']
    
    is_blog = any(ind in text_lower for ind in blog_indicators) or '/blog/' in url_lower or 'blog.' in url_lower
    is_product = any(ind in text_lower for ind in product_indicators) or '/product' in url_lower or '/pricing' in url_lower
    
    if is_blog and not is_product:
        return "blog"
    if is_product and not is_blog:
        return "product"
    return "general"

def extract_wordcount_analysis(html_content: str, url: str, target_keyword: Optional[str] = None) -> Dict[str, Any]:
    """
    Comprehensive wordcount analysis.
    """
    soup = BeautifulSoup(html_content, 'html.parser')
    
    # Total Word Count (all text except script/style/noscript)
    temp_soup = BeautifulSoup(html_content, 'html.parser')
    for el in temp_soup(["script", "style", "noscript"]):
        el.decompose()
    total_text = temp_soup.get_text()
    total_word_count = len(tokenize_words(total_text))
    
    # Visible Word Count
    visible_text = extract_visible_text(soup)
    visible_words = tokenize_words(visible_text)
    visible_word_count = len(visible_words)
    
    # Unique Word Count
    unique_words = set(tokenize_words(normalize_text(visible_text)))
    unique_word_count = len(unique_words)
    
    # Text-to-HTML Ratio
    html_size = len(html_content.encode('utf-8'))
    visible_text_size = len(visible_text.encode('utf-8'))
    ratio = (visible_text_size / html_size * 100) if html_size > 0 else 0
    
    # Sentence and Paragraph counts
    sentence_count = get_sentence_count(visible_text)
    paragraph_count = get_paragraph_count(soup)
    
    # Averages
    avg_sentence_len = round(visible_word_count / sentence_count, 2) if sentence_count > 0 else 0
    avg_paragraph_len = round(visible_word_count / paragraph_count, 2) if paragraph_count > 0 else 0
    
    # Keyword Density
    density = calculate_keyword_density(visible_text, target_keyword)
    
    # Thin Content
    thin_check = detect_thin_content(visible_word_count, unique_word_count)
    
    # Section Mapping
    section_mapping = extract_section_mapping(soup)
    section_breakdown = {k: round((v / visible_word_count * 100), 2) for k, v in section_mapping.items()} if visible_word_count > 0 else {}
    
    # Heading mapping (level prefixed)
    heading_mapping = {}
    for h in soup.find_all(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']):
        level = h.name.upper()
        h_text = h.get_text().strip()
        if h_text:
            key = f"{level}: {h_text}"
            # find words until next h of same or higher level
            words = []
            for s in h.next_siblings:
                if s.name and re.match(r'^h[1-6]$', s.name.lower()):
                    if int(s.name[1]) <= int(h.name[1]):
                        break
                if hasattr(s, 'get_text'):
                    words.append(s.get_text())
            heading_mapping[key] = len(tokenize_words(' '.join(words)))

    # Content Type Distribution
    content_type = detect_intent(visible_text, url)
    
    return {
        "totalWordCount": max(total_word_count, visible_word_count),
        "visibleWordCount": visible_word_count,
        "uniqueWordCount": unique_word_count,
        "textToHtmlRatio": round(ratio, 2),
        "sentenceCount": sentence_count,
        "paragraphCount": paragraph_count,
        "averageSentenceLength": avg_sentence_len,
        "averageParagraphLength": avg_paragraph_len,
        "keywordDensity": density,
        "thinContent": thin_check["thinContent"],
        "thinContentReason": thin_check["thinContentReason"],
        "duplicateContent": False, # Requires session
        "duplicateWithUrls": [], # Requires session
        "sectionWordCountMapping": section_mapping,
        "sectionWordCountBreakdown": section_breakdown,
        "headingWordCountMapping": heading_mapping,
        "wordCountDistribution": {
            "contentType": content_type,
            "isBlog": content_type == "blog",
            "isProduct": content_type == "product"
        }
    }
